File: ios/Classes/KlarnaSignInFlutterPlugin.swift
import Flutter
import UIKit
import KlarnaMobileSDK
import AuthenticationServices


public class KlarnaSignInFlutterPlugin: NSObject, FlutterPlugin, FlutterStreamHandler, KlarnaEventHandler, ASWebAuthenticationPresentationContextProviding {

    private var eventSink: FlutterEventSink?
    private var sdk: KlarnaSignInSDK?

    private var returnUrl: URL?
    private var environment: KlarnaEnvironment = .playground
    private var region: KlarnaRegion = .na
    private var theme: KlarnaTheme = .light

    public static func register(with registrar: FlutterPluginRegistrar) {
        let method = FlutterMethodChannel(name: "klarna_signin/methods", binaryMessenger: registrar.messenger())
        let events = FlutterEventChannel(name: "klarna_signin/events", binaryMessenger: registrar.messenger())

        let instance = KlarnaSignInFlutterPlugin()
        registrar.addMethodCallDelegate(instance, channel: method)
        events.setStreamHandler(instance)
    }

    // FlutterStreamHandler
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        return nil
    }
    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.eventSink = nil
        return nil
    }

    // Flutter Method Calls
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "initialize":
            guard let args = call.arguments as? [String: Any], let returnUrlStr = args["returnUrl"] as? String, let url = URL(string: returnUrlStr) else {
                result(FlutterError(code: "ARG_ERROR", message: "returnUrl missing", details: nil))
                return
            }
            returnUrl = url

            if let env = (args["environment"] as? String)?.lowercased() {
                environment = env == "production" ? .production : .playground
            }
            if let reg = (args["region"] as? String)?.uppercased() {
                switch reg { case "NA": region = .na; case "OC": region = .oc; default: region = .eu }
            }
            if let th = (args["theme"] as? String)?.lowercased() {
                switch th { case "light": theme = .light; case "dark": theme = .dark; default: theme = .automatic }
            }

            sdk = KlarnaSignInSDK(theme: theme, environment: environment, region: region, returnUrl: url, eventHandler: self)
            result(nil)

        case "signIn":
            guard let args = call.arguments as? [String: Any],
                  let clientId = args["clientId"] as? String,
                  let scope = args["scope"] as? String,
                  let market = args["market"] as? String else {
                result(FlutterError(code: "ARG_ERROR", message: "missing signIn args", details: nil))
                return
            }
            let locale = args["locale"] as? String

            guard let s = sdk else {
                result(FlutterError(code: "NOT_INITIALIZED", message: "Call initialize() first", details: nil))
                return
            }

            s.signIn(
                clientId: clientId,
                scope: scope,
                market: market,
                locale: locale,
                presentationContext: self
            )
            result(nil)

        default:
            result(FlutterMethodNotImplemented)
        }
         
    }

    // 递归序列化
    private func serializeParam(_ value: Any) -> Any {
        if let token = value as? KlarnaSignInToken {
            return [
                "idToken": token.idToken ?? "",
                "accessToken": token.accessToken ?? "",
                "refreshToken": token.refreshToken ?? "",
                "scope": token.scope ?? "",
                "tokenType": token.tokenType ?? "",
                "expiresIn": token.expiresIn ?? 0
            ]
        } else if let dict = value as? [String: Any] {
            var serialized: [String: Any] = [:]
            for (k, v) in dict {
                serialized[k] = serializeParam(v)
            }
            return serialized
        } else if let array = value as? [Any] {
            return array.map { serializeParam($0) }
        } else {
            return value
        }
    }
    // KlarnaEventHandler
    public func klarnaComponent(_ klarnaComponent: KlarnaComponent, dispatchedEvent event: KlarnaProductEvent) {
        
    let map: [String: Any] = [
        "action": event.action,
        "params": serializeParam(event.params)
    ]
            // 确保在主线程回调 Flutter
    DispatchQueue.main.async {
        self.eventSink?(map)
    }
        
    }
    public func klarnaComponent(_ klarnaComponent: KlarnaComponent, encounteredError error: KlarnaError) {
        
        var map: [String: Any] = ["action": "ERROR"]
        map["params"] = ["message": error.localizedDescription]
            // 确保在主线程回调 Flutter
    DispatchQueue.main.async {
        self.eventSink?(map)
    }
        
    }
    // ASWebAuthenticationPresentationContextProviding
    public func presentationAnchor(for session: ASWebAuthenticationSession) -> ASPresentationAnchor {
        if Thread.isMainThread {
            if let windowScene = UIApplication.shared.connectedScenes.first as? UIWindowScene,
               let key = windowScene.windows.first(where: { $0.isKeyWindow }) {
                return key
            } else {
                return UIApplication.shared.windows.first ?? UIWindow()
            }
        } else {
            var anchor: ASPresentationAnchor? = nil
            DispatchQueue.main.sync {
                if let windowScene = UIApplication.shared.connectedScenes.first as? UIWindowScene,
                   let key = windowScene.windows.first(where: { $0.isKeyWindow }) {
                    anchor = key
                } else {
                    anchor = UIApplication.shared.windows.first
                }
            }
            return anchor ?? UIWindow()
        }
    }
}

File: android/src/main/kotlin/com/example/klarna_sign_in_flutter/KlarnaSignInFlutterPlugin.kt

package com.example.klarna_sign_in_flutter

import android.app.Activity
import android.content.Context
import androidx.annotation.MainThread
import com.klarna.mobile.sdk.api.KlarnaEventHandler
import com.klarna.mobile.sdk.api.KlarnaProductEvent
import com.klarna.mobile.sdk.api.KlarnaEnvironment
import com.klarna.mobile.sdk.api.KlarnaRegion
import com.klarna.mobile.sdk.api.KlarnaTheme
import com.klarna.mobile.sdk.api.signin.KlarnaSignInSDK
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import com.google.gson.Gson
class KlarnaSignInFlutterPlugin : FlutterPlugin, MethodChannel.MethodCallHandler,
    EventChannel.StreamHandler, ActivityAware ,KlarnaEventHandler {

    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private var eventSink: EventChannel.EventSink? = null

    private var activity: Activity? = null
    private var context: Context? = null

    private var returnUrl: String? = null
    private var environment: KlarnaEnvironment = KlarnaEnvironment.PLAYGROUND
    private var region: KlarnaRegion = KlarnaRegion.EU
    private var theme: KlarnaTheme = KlarnaTheme.AUTOMATIC

    private var sdk: KlarnaSignInSDK? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel = MethodChannel(binding.binaryMessenger, "klarna_signin/methods")
        methodChannel.setMethodCallHandler(this)
        eventChannel = EventChannel(binding.binaryMessenger, "klarna_signin/events")
        eventChannel.setStreamHandler(this)
        context = binding.applicationContext
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        context = null
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
    }
    override fun onDetachedFromActivityForConfigChanges() { activity = null }
    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) { activity = binding.activity }
    override fun onDetachedFromActivity() { activity = null }

    // EventChannel.StreamHandler
    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) { eventSink = events }
    override fun onCancel(arguments: Any?) { eventSink = null }

    // MethodChannel
    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
          when (call.method) {
            "initialize" -> {
                val args = call.arguments as Map<*, *>
                returnUrl = args["returnUrl"] as String
                environment = when ((args["environment"] as String?)?.lowercase()) {
                    "production" -> KlarnaEnvironment.PRODUCTION
                    else -> KlarnaEnvironment.PLAYGROUND
                }
                region = when ((args["region"] as String?)?.uppercase()) {
                    "NA" -> KlarnaRegion.NA
                    "OC" -> KlarnaRegion.OC
                    else -> KlarnaRegion.EU
                }
                theme = when ((args["theme"] as String?)?.lowercase()) {
                    "light" -> KlarnaTheme.LIGHT
                    "dark" -> KlarnaTheme.DARK
                    else -> KlarnaTheme.AUTOMATIC
                }

                val act = activity
                val ret = returnUrl
                if (act != null && ret != null) {
                    sdk = KlarnaSignInSDK(
                        activity = act,
                        returnURL = ret,
                        eventHandler = this,
                        environment = environment,
                        region = region,
                        theme = theme
                    )
                    result.success(null)
                } else {
                    result.error("NO_ACTIVITY_OR_URL", "Activity or returnUrl missing", null)
                }
            }
            "signIn" -> {
                val args = call.arguments as Map<*, *>
                val clientId = args["clientId"] as String
                val scope = args["scope"] as String
                val market = args["market"] as String
                val locale = args["locale"] as String?

                val s = sdk
                if (s == null) {
                    result.error("NOT_INITIALIZED", "Call initialize() first", null)
                    return
                }

                s.signIn(
                    clientId = clientId,
                    scope = scope,
                    market = market,
                    locale = locale
                )
                result.success(null)
            }
            else -> result.notImplemented()
        }
    }


    @MainThread
    override fun onEvent(klarnaComponent: com.klarna.mobile.sdk.api.component.KlarnaComponent, event: com.klarna.mobile.sdk.api.KlarnaProductEvent) {
        val map = HashMap<String, Any?>()
        val gson = Gson()
        val paramsMap: HashMap<String, Any?> = gson.fromJson(
            gson.toJson(event.params),
            HashMap::class.java
        ) as HashMap<String, Any?>

        map["action"] = event.action
        map["params"] = paramsMap
        eventSink?.success(map)
    }   
    @MainThread
    override fun onError(klarnaComponent: com.klarna.mobile.sdk.api.component.KlarnaComponent, error: com.klarna.mobile.sdk.KlarnaMobileSDKError) {
        val map = HashMap<String, Any?>()
        map["action"] = "ERROR"
        map["params"] = mapOf("message" to (error.message ?: error.toString()))
        eventSink?.success(map)
    }
}